Storage/SQL/BrowserDB.swift

```diff
@@ -263,25 +265,43 @@ extension BrowserDB {
     }
 
     func run(sql: String, withArgs args: Args? = nil) -> Success {
-        var err: NSError?
-        return self.withWritableConnection(&err) { (conn, inout err: NSError?) -> Success in
-            err = conn.executeChange(sql, withArgs: args)
-            if err == nil {
+        let deferred = Deferred<Result<()>>()
+
+        dispatch_async(queue) {
+            let start = NSDate()
+            var err: NSError?
+            let result = self.withWritableConnection(&err) { (conn, inout err: NSError?) -> NSError? in
+                err = conn.executeChange(sql, withArgs: args)
                 log.debug("Modified rows: \(conn.numberOfRowsModified).")
-                return succeed()
+                return err
+            }
+
+            if err == nil {
+                deferred.fill(Result(success: ()))
+            } else {
+                deferred.fill(Result(failure: DatabaseError(err: err)))
             }
-            return deferResult(DatabaseError(err: err))
+            let now = NSDate()
+            log.debug("Query \(sql) took \(now.timeIntervalSinceDate(start))")
         }
+
+        return deferred
     }
 
     func runQuery<T>(sql: String, args: Args?, factory: SDRow -> T) -> Deferred<Result<Cursor<T>>> {
-        func f(conn: SQLiteDBConnection, inout err: NSError?) -> Cursor<T> {
-            return conn.executeQuery(sql, factory: factory, withArgs: args)
-        }
+        let result = Deferred<Result<Cursor<T>>>()
 
-        var err: NSError? = nil
-        let cursor = self.withReadableConnection(&err, callback: f)
+        dispatch_async(queue) {
+            let start = NSDate()
+            var err: NSError? = nil
+            let cursor = self.withReadableConnection(&err, callback: { (conn: SQLiteDBConnection, inout err: NSError?) -> Cursor<T> in
```

Comment:
Can we share some of this with `run`, in a generic function? I didn't reuse code between them because there wasn't much boilerplate, but now there's a little more.